lib/spack/spack/cmd/config.py

```diff
@@ -431,6 +443,79 @@ def config_revert(args):
         tty.msg(msg.format(cfg_file))
 
 
+def config_prefer_upstream(args):
+    """Generate a packages config based on the configuration of all upstream
+    installs."""
+
+    scope = args.scope
+    if scope is None:
+        scope = spack.config.default_modify_scope('packages')
+
+    all_specs = set(spack.store.db.query(installed=True))
+    local_specs = set(spack.store.db.query_local(installed=True))
+    pref_specs = local_specs if args.local else all_specs - local_specs
+
+    conflicting_variants = set()
+
+    pkgs = {}
+    for spec in pref_specs:
+        # Collect all the upstream compilers and versions for this package.
+        pkg = pkgs.get(spec.name, {
+            'version': [],
+            'compiler': [],
+        })
+        pkgs[spec.name] = pkg
+
+        # We have no existing variant if this is our first added version.
+        existing_variants = pkg.get('variants',
+                                    None if not pkg['version'] else '')
```

Comment:
```suggestion
                                    None)
```
Below you check whether it's None, not whether it's False-ish

Reply:
This is intentional. None denotes that this is the first time we've seen this package, while an empty string is simply the lack of variants if there weren't any.

Reply:
Intentionally conflicting no variants specified with any specified variants?

hmm... so we tried none-specified as if they specifically requested all the defaults... I guess that works well for this use case, just not what we usually do.

Reply:
Ok, then I see the other way to get rid of the test failure below. I'll suggest it inline.

Reply:
Yes, the variants string would be an empty string if no variants differed from the defaults. 
The goal is to make a concise configuration, not an exhaustive one.